miden-crypto/benches/merkle.rs

```diff
@@ -7,45 +7,220 @@
 //! `benches/smt-subtree.rs`.
 use std::{hint, time::Duration};
 
-use criterion::{BatchSize, Criterion, criterion_group, criterion_main};
-use miden_crypto::{Felt, ONE, Word, merkle::MerkleTree};
+use criterion::{BatchSize, Bencher, Criterion, criterion_group, criterion_main};
+use miden_crypto::{
+    Felt, ONE, Word,
+    merkle::{MerklePath, MerkleTree, NodeIndex},
+};
 
 mod common;
 use common::data::*;
 
-fn balanced_merkle_even(c: &mut Criterion) {
-    c.bench_function("balanced-merkle-even", |b| {
+benchmark_multi!(
+    balanced_merkle_even,
+    "balanced-merkle-even",
+    &[4, 8, 16, 32, 64, 128, 256],
+    |b: &mut Bencher<'_>, num_leaves: &usize| {
         b.iter_batched(
             || {
-                let entries: Vec<Word> =
-                    (0..256).map(|i| Word::new([Felt::new(i), ONE, ONE, Felt::new(i)])).collect();
-                assert_eq!(entries.len(), 256);
+                let entries: Vec<Word> = (0..*num_leaves)
+                    .map(|i| Word::new([Felt::new(i as u64), ONE, ONE, Felt::new(i as u64)]))
+                    .collect();
+                assert_eq!(entries.len(), *num_leaves);
                 entries
             },
             |leaves| {
                 let tree = MerkleTree::new(hint::black_box(leaves)).unwrap();
-                assert_eq!(tree.depth(), 8);
+                assert_eq!(tree.depth(), num_leaves.ilog2() as u8);
             },
             BatchSize::SmallInput,
         );
-    });
-}
+    }
+);
 
-fn balanced_merkle_rand(c: &mut Criterion) {
-    let mut seed = [0u8; 32];
-    c.bench_function("balanced-merkle-rand", |b| {
+benchmark_multi!(
+    balanced_merkle_rand,
+    "balanced-merkle-rand",
+    &[4, 8, 16, 32, 64, 128, 256],
+    |b: &mut Bencher<'_>, num_leaves: &usize| {
+        let mut seed = [0u8; 32];
         b.iter_batched(
             || {
-                let entries: Vec<Word> = (0..256).map(|_| generate_word(&mut seed)).collect();
-                assert_eq!(entries.len(), 256);
+                let entries: Vec<Word> =
+                    (0..*num_leaves).map(|_| generate_word(&mut seed)).collect();
+                assert_eq!(entries.len(), *num_leaves);
                 entries
             },
             |leaves| {
                 let tree = MerkleTree::new(hint::black_box(leaves)).unwrap();
-                assert_eq!(tree.depth(), 8);
+                assert_eq!(tree.depth(), num_leaves.ilog2() as u8);
             },
             BatchSize::SmallInput,
         );
+    }
+);
+
+// MERKLE TREE BENCHMARKS
+// ================================================================================================
+
+benchmark_with_setup_data!(
+    merkle_tree_root,
+    Duration::from_secs(10),
+    10,
+    "merkle-tree-root",
+    || {
+        let entries: Vec<Word> =
+            (0..256).map(|i| Word::new([Felt::new(i), ONE, ONE, Felt::new(i)])).collect();
```

Comment:
nit: seems like this pattern is repeated many times throughout, maybe worth extracting it for reuse

Reply:
Also, why this specific pattern and not variations of it